btc-types/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+mod u256;
+use borsh::{BorshDeserialize, BorshSerialize};
+use serde::{Deserialize, Serialize};
+
+pub use u256::U256;
+pub type Target = U256;
+pub type Work = U256;
+
+#[derive(
+    BorshDeserialize,
+    BorshSerialize,
+    Serialize,
+    Deserialize,
+    Clone,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Debug,
+    Default,
+)]
+pub struct H256(pub [u8; 32]);
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize, Clone, Debug)]
+pub struct Header {
+    /// Block version, now repurposed for soft fork signalling.
+    pub version: i32,
+    /// Reference to the previous block in the chain.
+    pub prev_block_hash: H256,
+    /// The root hash of the merkle tree of transactions in the block.
+    pub merkle_root: H256,
+    /// The timestamp of the block, as claimed by the miner.
+    pub time: u32,
+    /// The target value below which the blockhash must lie.
+    pub bits: u32,
+    /// The nonce, selected to obtain a low enough blockhash.
+    pub nonce: u32,
+}
+
+impl Header {
+    /// The number of bytes that the block header contributes to the size of a block.
+    // Serialized length of fields (version, prev_blockhash, merkle_root, time, bits, nonce)
+    pub const SIZE: usize = 4 + 32 + 32 + 4 + 4 + 4; // 80
+
+    /// Computes the target (range [0, T] inclusive) that a blockhash must land in to be valid.
+    pub fn target(&self) -> Target {
+        // This is a floating-point "compact" encoding originally used by
+        // OpenSSL, which satoshi put into consensus code, so we're stuck
+        // with it. The exponent needs to have 3 subtracted from it, hence
+        // this goofy decoding code. 3 is due to 3 bytes in the mantissa.
+        let (mant, expt) = {
+            let unshifted_expt = self.bits >> 24;
+            if unshifted_expt <= 3 {
+                (
+                    (self.bits & 0xFFFFFF) >> (8 * (3 - unshifted_expt as usize)),
+                    0,
+                )
+            } else {
+                (self.bits & 0xFFFFFF, 8 * ((self.bits >> 24) - 3))
+            }
+        };
+
+        // The mantissa is signed but may not be negative.
+        if mant > 0x7F_FFFF {
+            Target::ZERO
+        } else {
+            U256::from(mant) << expt
```

Comment:
should we check that `expt <= 29*8`?

Reply:
@olga24912 Where did this number come from?

Reply:
We have max 256 bit = 32 bytes (U256). mant = 3 bytes, so we can max shift on 29 bytes (32 - 3 bytes), otherwise we cut mant.